### PR TITLE
Fat 1002

### PR DESCRIPTION
**Pull request by yusuf-murodov-epam**
## Purpose
[FAT-1002](https://issues.folio.org/browse/FAT-1002)

## Approach
Created all the instances, i.e postInstance, postLocation, postItems and so on...
Check out created items for the user
Then matched loans size for the user, and other attributes to see whether they are valid loan

Tested get circulation loans endpoint parameters by Scenario outlines with positive and negative test cases

### CODE REVIEW

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 1]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
+    * def extItemId1 = call uuid1
+    * def extItemId2 = call uuid1
+    * def limit = 5
+
+    #post items
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostInstance') { extInstanceTypeId: #(extInstanceTypeId) }
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostServicePoint')
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostLocation') { extInstitutionId: #(extInstitutionId), extCampusId: #(extCampusId), extLibraryId: #(extLibraryId) }
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostHoldings')
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostItem') { extItemBarcode: '989898', extItemId: #(extItemId1) }
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:101]: you may use * def extItemBarcode = random(100000) to avoid hard barcode

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 2]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
     And match response.itemId == itemId
+
+  Scenario: Get Loans collection
+    * def extInstanceTypeId = call uuid1
+    * def extInstitutionId = call uuid1
+    * def extCampusId = call uuid1
+    * def extLibraryId = call uuid1
+    * def extItemId1 = call uuid1
+    * def extItemId2 = call uuid1
+    * def limit = 5
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:94]: why do we need it ?

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 3]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
+
+    # checkOut the first item
+    * def checkOutResponse1 = call read('classpath:domain/mod-circulation/features/util/initData.feature@PostCheckOut') { extCheckOutUserBarcode: '121212', extCheckOutItemBarcode: '989898' }
+    # checkOut the second item
+    * def checkOutResponse2 = call read('classpath:domain/mod-circulation/features/util/initData.feature@PostCheckOut') { extCheckOutUserBarcode: '121212', extCheckOutItemBarcode: '989899' }
+
+    # Get a paged collection of patron
+    Given path 'circulation/loans'
+    And param query = '(userId=="' + userId + '")'
+    And param limit = limit
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:115]: why do we need it ?

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 4]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
+    Then status 200
+
+    Examples:
+      | id          | limit       |
+      | userId      | 1           |
+      | userId      | 0           |
+
+  # ================= negative test cases for circulation loans collection =================
+
+  Scenario Outline: Get Loans collection
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:140]: I believe we don't need it

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 5]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
+    Then status 200
+    And match response.totalRecords == 2
+    And match response.loans[0].id == checkOutResponse1.response.id
+    And match response.loans[0].status.name == 'Open'
+    And match response.loans[1].id == checkOutResponse2.response.id
+    And match response.loans[1].status.name == 'Open'
+
+    # ================= positive test cases for circulation loans collection =================
+
+  Scenario Outline: Get Loans collection
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:126]: I believe we don't need it

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 6]
```diff
@@ -83,3 +83,72 @@ Feature: Loans tests
...
+    Given path 'circulation/loans'
+    And param query = '(userId=="' + userId + '")'
+    And param limit = limit
+    When method GET
+    Then status 200
+    And match response.totalRecords == 2
+    And match response.loans[0].id == checkOutResponse1.response.id
+    And match response.loans[0].status.name == 'Open'
+    And match response.loans[1].id == checkOutResponse2.response.id
+    And match response.loans[1].status.name == 'Open'
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:122]:
```suggestion
    And match response == { totalRecords: #present, loans: #present }
    And match response.totalRecords == 2
    And match response.loans[0] == { lostItemPolicyId: #present, metadata: #present, item: #present, dueDate: #present, checkoutServicePointId: #present, borrower: #present, feesAndFines: #present, userId: #present, patronGroupAtCheckout: #present, overdueFinePolicy: #present, checkoutServicePoint: #present, itemId: #present, loanPolicyId: #present, itemEffectiveLocationIdAtCheckOut: #present, loanDate: #present, action: #present, overdueFinePolicyId: #present, lostItemPolicy: #present, id: #present, loanPolicy: #present, status: #present }
```

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 2 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:94]: extItemId1 and extItemId2 are needed to differentiate our items, since I am checking out 2 different items for one user so that I can see more items on the response, if it is superfluous action, I can just do with one item, then these definitions can also be removed

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 3 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:115]: When I reproduce on the UI firstly, I noticed the end point has some query parameters, so I also provided on this test as well

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 7]
```diff
@@ -85,70 +85,31 @@ Feature: Loans tests
...
-    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostInstance') { extInstanceTypeId: #(extInstanceTypeId) }
+    #post an item
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostInstance')
     * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostServicePoint')
-    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostLocation') { extInstitutionId: #(extInstitutionId), extCampusId: #(extCampusId), extLibraryId: #(extLibraryId) }
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostLocation')
     * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostHoldings')
-    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostItem') { extItemBarcode: '989898', extItemId: #(extItemId1) }
-    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostItem') { extItemBarcode: '989899', extItemId: #(extItemId2) }
+    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostItem') { extItemBarcode: #(extItemBarcode) }
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:98]: to create two items was a good idea, why did you remove one of them I haven't clue

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 3 above)
Comment (viktor-draban) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:109]: please rid of from 'limit'. Honestly this endpoint has more query parameters(you can see all of them on https://s3.amazonaws.com/foliodocs/api/mod-circulation/circulation.html#circulation_loans_get) but all of them aren't obligatory and in our case we don't need them

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 8]
```diff
@@ -85,70 +85,31 @@ Feature: Loans tests
...
-    * def extItemId2 = call uuid1
-    * def limit = 5
+    
+    * def limit = random(2000)
+    * def extItemBarcode = random(10000)
+    * def extUserBarcode = random(10000)
 
-    #post items
-    * call read('classpath:domain/mod-circulation/features/util/initData.feature@PostInstance') { extInstanceTypeId: #(extInstanceTypeId) }
+    #post an item
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:93]:
```suggestion
    # post an item
```

Comment (viktor-draban) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:93]: and if you return two items --->  # post items

File: mod-circulation/src/main/resources/domain/mod-circulation/features/util/initData.feature [hunk 9]
```diff
@@ -101,6 +101,7 @@ Feature: init data for mod-circulation
   Scenario: create item
     * def permanentLoanTypeId = call uuid1
     * def intMaterialTypeId = call uuid1
+    * def intEntityId = call uuid1
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/util/initData.feature:104]:
```suggestion
    * def intItemId = call uuid1
```

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 8 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature]: ok, I returned two items and fixed

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 5 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature]: removed it)

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 4 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature]: removed

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 10]
```diff
@@ -83,3 +83,158 @@ Feature: Loans tests
...
+    And match response.loans[0].id == checkOutResponse1.response.id
+    And match response.loans[1] == { id: #present, lostItemPolicyId: #present, metadata: #present, item: #present, dueDate: #present, checkoutServicePointId: #present, borrower: #present, feesAndFines: #present, userId: #present, patronGroupAtCheckout: #present, overdueFinePolicy: #present, checkoutServicePoint: #present, itemId: #present, loanPolicyId: #present, itemEffectiveLocationIdAtCheckOut: #present, loanDate: #present, action: #present, overdueFinePolicyId: #present, lostItemPolicy: #present, id: #present, loanPolicy: #present, status: #present }
+    And match response.loans[1].id == checkOutResponse2.response.id
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:239]: remove

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature [hunk 11]
```diff
@@ -83,3 +83,158 @@ Feature: Loans tests
     Then status 200
     And match response.itemStatusPriorToCheckIn == 'Checked out'
     And match response.itemId == itemId
+
+  Scenario: Get Loans collection
```
Comment (viktor-draban) [inline mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:87]:
```suggestion
  Scenario: When get loans for a patron is called, return a paged collection of loans for that patron with all data as specified in the circulation/loans API
```

rename plz

**Comment by viktor-draban** [comment]
image
fix it

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 10 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:239]: What should I remove in this section?

File: mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature (hunk 11 above)
Comment (yusuf-murodov-epam) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:87]: OK, Every time we should name our scenario similar to ticket name, right?

Comment (viktor-draban) [reply mod-circulation/src/main/resources/domain/mod-circulation/features/loans.feature:87]: At least while we write karate tests based on the documentation
